### ISSUE TITLE
[RFC]: Hidden states processor

### ISSUE THREAD

**Issue by DarkLight1337** [vllm-project/vllm#12249]
### Motivation.

Since #10674, vLLM uses Pooler to extract hidden states from the model and convert them to embeddings, class probabilities, and so on. However, this is still not user-friendly enough:

- We have separate model runners for generative and pooling models. This complicates the effort to return hidden states alongside generated text (e.g.: #6165, #11397, #11577, #11606, #11905)
- Setting the default Pooler based on downstream task only covers the common cases. It may be required to use `--override-pooler-config` which isn't that intuitive to use (e.g. #12085). Even so, we still lack support for custom processing of hidden states (e.g. #11065, #11881, #12162)

### Proposed Change.

Similar to `LogitsProcessor` (#1469), we can pass a custom `HiddenStatesProcessor` in `SamplingParams` and `PoolingParams` to postprocess the hidden states and return them in the output. This provides maximum flexibility and enables the same model to be used for different downstream tasks.

The interface of `HiddenStatesProcessor` is similar to `VllmModelForTextGeneration.compute_logits` and `VllmModelForPooling.pooler`:

```py
H = TypeVar("H", default=torch.Tensor)
R = TypeVar("R", default=torch.Tensor)

class HiddenStatesProcessor(Protocol[H, R]):
    def __call__(self, model: VllmModel[H], hidden_states: H) -> R:
        ...
```

The default poolers for each downstream task will be implemented as built-in `HiddenStatesProcessor` classes.

- `IdentityHiddenStatesProcessor`: Returns hidden states directly. (mainly for reward models)
- `NormalizeHiddenStatesProcessor`: Applies normalization to hidden states. (mainly for prompt embedding)
- `SoftmaxHiddenStatesProcessor`: Applies softmax to hidden states. (mainly for classification)
- `StepHiddenStatesProcessor`: Applies step processing to hidden states (mainly for PRM models)

The existing pooling APIs (`LLM.encode`, `LLM.embed`, `LLM.score`, etc.) will be updated to use these `HiddenStatesProcessor` automatically.

To get logits from the hidden states, we can have a new hidden states processor that references the LM head of the model. For example:

```py
class ExtractLogitsHiddenStatesProcessor(HiddenStatesProcessor):
    def __init__(self, lm_head_name: str = "lm_head") -> None:
        self.lm_head_name = lm_head_name

    def __call__(self, model: VllmModel, hidden_states: torch.Tensor) -> torch.Tensor:
        lm_head = getattr(model, self.lm_head_name)
        assert isinstance(lm_head, VocabParallelEmbedding)

        logits = lm_head.linear_method.apply(lm_head, hidden_states)
        return logits
```

We can also have the model generate multi-modal outputs:

```py
class ImageFromHiddenStatesProcessor(HiddenStatesProcessor[torch.Tensor, list[Image]]):
    def __init__(self, decoder_name: str = "image_decoder") -> None:
        self.decoder_name = decoder_name
        self._to_pil_image = torchvision.transforms.v2.ToPILImage()

    # Suppose hidden_states is the output of the model's encoder without calling the vision decoder
    def __call__(self, model: VllmModel, hidden_states: torch.Tensor) -> list[Image]:
        image_decoder = getattr(model, self.decoder_name)
        images = image_decoder(hidden_states)  # Shape: [N, C, H, W]
        return [self._to_pil_image(image) for image in images.cpu()]
```

### Feedback Period.

Around 2 weeks? See when I have time to work on this...

### CC List.

@simon-mo @youkaichao @robertgshaw2-redhat @ywang96 @Isotr0py @maxdebayser @flaviabeo @HwwwwwwwH 

### Any Other Things.

Since the regular model runner can also return hidden states, we should consider merging the functionality of `PoolingModelRunner` with the regular `ModelRunner` in V1 (#8779) to simplify our codebase. I think the only difference is that `PoolingModelRunner` uses dummy KV caches?

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by youkaichao** [comment]
> we can pass a custom HiddenStatesProcessor in SamplingParams

can we pass it as an argument to `LLM`? I don't think people use different processor for each request. It should be the same across one inference instance.

**Comment by DarkLight1337** [comment]
> can we pass it as an argument to LLM? I don't think people use different processor for each request. It should be the same across one inference instance.

Some users want to use the same LLM engine for online generation and embedding, but not necessarily both in the same request. It would be a waste of resources to run both in that case.

**Comment by youkaichao** [comment]
I don't think we need to support that. One engine should do one task, otherwise the code would be super-complicated, and would be difficult to optimize.

**Comment by youkaichao** [comment]
complicated sampling parameter is a major factor why vllm became slower previously. We should be very careful about runtime cost that happens per-request. While generally I'm fine with adding engine-level features that do not affect per-request performance.

**Comment by DarkLight1337** [comment]
I understand your concern. Keep in mind though that hidden states processors are optional and should not affect performance unless they are being used. Unless I'm mistaken, our Poolers aren't being optimized either, so the performance should be similar as now.

**Comment by youkaichao** [comment]
> Keep in mind though that hidden states processors are optional and should not affect performance except when they are being used.

adding sampling parameters will in general slow down the inference, because sometimes we need to pass the object across process.

I think hidden states processors should only be instance-level, users can only specify one processor for one instance.

**Comment by DarkLight1337** [comment]
> > We can pass a dictionary of "allowed" hidden states processors to LLM at startup time so vLLM can profile and optimize them. Then at inference time, we only let the user select from these processors.
> 
> that would be even worse, since you need to validate the processor during runtime, per-request.

I mean that the user only passes the (string) key of the processor in the initial dictionary without sending the actual object.

**Comment by comaniac** [comment]
I'm also in favor of setting it at the startup time so that we could better profile and avoid OOM. In general allowing one endpoint to support various pooling mechanism seems not that common. We could mark it as a limitation for now, and think about improvements in the future if there are high demands.

**Comment by youkaichao** [comment]
> OOM

out-of-memory error is also a valid concern. if the engine can serve multiple types of requests, then the memory profiling stage would be super complicated, and I doubt if it is even possible. under high load, lots of edge case can occur.

there's also scheduling challenges, the memory cost of generation and embedding can require different memory, and one single token budget would not be enough to quantify and bound the memory usage.

**Comment by maxdebayser** [comment]
In the past there were problems with `LogitsProcessors` that can't be pickled as they were passed to the `LLMEngine`. Setting up a single `HiddenStatesProcessor` at startup time would also avoid that issue.

**Comment by bhaktatejas922** [comment]
This still planned on going in, or has it been deprioritized for now?

**Comment by christian-pinto** [comment]
@DarkLight1337, @maxdebayser has there been progress on this? I am now in the position to start working on this in the context of https://github.com/vllm-project/vllm/issues/16052

My main focus would be on embedding type models, but this feature should be supported for any model as to providing a way of generating different output types.

I could start working on it focusing on V0 and move to V1 when it is stable enough?

**Comment by DarkLight1337** [comment]
Let's land the V1 embedding models PR first before working on this. @WoosukKwon requested that we make a separate model runner for now for pooling models.

**Comment by maxdebayser** [comment]
Yes, with the hidden states processor one of the main concerns expressed above is that it would make it difficult to have a warmup logic for memory profiling, cuda graphs and so on. In my original V1 embedding PR this problem is exacerbated by the overlap with the generative model logic. By having a separate model runner for embedding models, we sacrifice the possibility to serve pooling and sampling requests with the same model, but it should simplify things a lot. Especially because pooling models have no decode phase, the memory usage for processing the hidden states after the inference has been run should be less concerning, so perhaps we can get away with profiling only this part.

In parallel to the V1 embedding support, I've started thinking about the hidden states processor. Currently a dilemma I'm facing is this: in the language models all prompts of the batch are concatenated in the output instead of having a batch dimension and padding in the prompt dimension. Some of the pooling algorithms can be applied directly on the concatenated prompts, thus being more efficient, and some require extracting the hidden states for each prompt first. To implement custom hidden states processors, the easiest would be to always extract the prompts first, but it could lead to lower performance.

**Comment by maxdebayser** [comment]
The PR for a dedicated embedding model runner is now ready for review: https://github.com/vllm-project/vllm/pull/18015 .
Currently there are several merge conflicts due to the KVCache groups PR that was reverted: https://github.com/vllm-project/vllm/pull/18459 . But since it seems that this is only temporary, I'm waiting a bit before trying to fix the conflicts.

In this PR there is support for prefix caching and chunked prefill for decoder embedding model if the pooler supports it. "LAST" pooling where the last hidden state is used to calculate the embedding vector is currently the only supported one. To support other pooling types such as MEAN, the pooler has to become stateful to accumulate the current embedding vector, in the same way as the prompt logits are handled in the sampling model runner. So perhaps this is another requirement for the Hidden States Processer.

Another change that I did in this PR is to extract the hidden states per request in the model runner. This means that the pooler always gets a list of tensors. This will make it easier to design a common interface for the hidden state processors.

**Comment by DarkLight1337** [comment]
Closing as completed by https://github.com/vllm-project/vllm/pull/22820
